js/language_server/Main.ml

```diff
@@ -0,0 +1,105 @@
+(* Austin Theriault
+ *
+ * Copyright (C) 2019-2023 Semgrep, Inc.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * version 2.1 as published by the Free Software Foundation, with the
+ * special exception on linking described in file LICENSE.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file
+ * LICENSE for more details.
+ *)
+
+(* Commentary *)
+(*  *)
+
+(*****************************************************************************)
+(* Prelude *)
+(*****************************************************************************)
+
+open Js_of_ocaml
+open Semgrep_js_shared
+
+let read_line_ref = ref (fun () -> Lwt.return None)
+let read_exactly_ref = ref (fun _ -> Lwt.return None)
+
+module Io = RPC_server.MakeLSIO (struct
+  type input = unit
+  type output = unit
+
+  let read_line _ = !read_line_ref ()
+  let stdin = ()
+  let stdout = ()
+  let flush _ = Lwt.return ()
+  let atomic f oc = f oc
+
+  let write _ str =
+    (* nosem *)
+    print_string str;
+    Lwt.return ()
+
+  let read_exactly _ n = !read_exactly_ref n
+end)
+
+(*****************************************************************************)
+(* Code *)
+(*****************************************************************************)
+let _Promise = Js.Unsafe.global##._Promise
+let ( let+ ) = Lwt.( >|= )
+
+let promise_of_lwt lwt =
+  new%js _Promise
+    (Js.wrap_callback (fun resolve reject ->
+         try%lwt
+           let+ res = lwt () in
+           Js.Unsafe.fun_call resolve [| Js.Unsafe.inject res |]
+         with
+         | e ->
+             let msg = Printexc.to_string e in
+             Firebug.console##error (Js.string msg);
+             Js.Unsafe.fun_call reject
+               [| Js.Unsafe.inject (new%js Js.error_constr (Js.string msg)) |]))
+
+(* Stolen from Logs' logs_browser.ml *)
+let ppf, flush =
+  let b = Buffer.create 255 in
```

Comment:
Is the 255 here a limit for printing? Would be nice to have this be in a descriptively named variable with a comment.

Reply:
I'll be honest, not sure, just stole it from the Logs library

Reply:
There's already a comment explaining that tho